src/include/btree_cmp_inline.h

```diff
@@ -166,6 +300,143 @@ __wt_compare_bounds(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_ITEM *key, u
+     */
+    u = _mm_loadu_si128((const __m128i *)(uendp - WT_VECTOR_SIZE));
+    t = _mm_loadu_si128((const __m128i *)(tendp - WT_VECTOR_SIZE));
+    match = (size_t)(uendp - ustartp) - WT_VECTOR_SIZE;
+
+final128:
+    ufirst = (uint64_t)_mm_extract_epi64(u, 0);
+    tfirst = (uint64_t)_mm_extract_epi64(t, 0);
+    firsteq = ufirst == tfirst;
+    u64 = firsteq ? (uint64_t)_mm_extract_epi64(u, 1) : ufirst;
+    t64 = firsteq ? (uint64_t)_mm_extract_epi64(t, 1) : tfirst;
+    match += firsteq ? sizeof(uint64_t) : 0;
+
+#ifndef WORDS_BIGENDIAN
+    u64 = __wt_bswap64(u64);
+    t64 = __wt_bswap64(t64);
+#endif
+
+    WT_LEADING_ZEROS(u64 ^ t64, leading_zero_bytes);
+    match += (size_t)leading_zero_bytes;
+    *matchp = match;
+
+    return (u64 < t64 ? -1 : u64 > t64 ? 1 : lencmp);
+}
+#else
+/*
+ * __lex_compare_skip_gt_16 --
+ *     Lexicographic comparison routine for data greater than 16 bytes, skipping leading bytes.
+ *     Returns: < 0 if user_item is lexicographically < tree_item = 0 if user_item is
+ *     lexicographically = tree_item > 0 if user_item is lexicographically > tree_item We use the
+ *     names "user" and "tree" so it's clear in the btree code which the application is looking at
+ *     when we call its comparison function. Some platforms like ARM offer dedicated instructions
+ *     for reading 16 bytes at a time, allowing for faster comparisons. Some platforms like ARM
+ *     offer dedicated instructions for reading 16 bytes at a time, allowing for faster comparisons.
+ */
+static WT_INLINE int
+__lex_compare_skip_gt_16(
+  const uint8_t *ustartp, const uint8_t *tstartp, size_t len, int lencmp, size_t *matchp)
+{
+    struct {
+        uint64_t a, b;
+    } tdata, udata;
+    size_t match;
+    uint64_t t64, u64;
+    int leading_zero_bytes;
+    const uint8_t *tendp, *treep, *uendp, *userp;
+    bool firsteq;
+
+    match = *matchp;
+    uendp = ustartp + len;
+    tendp = tstartp + len;
+
+    /*
+     * Compare 16 bytes at a time until we find a difference or run out of 16 byte chunks to
+     * compare.
+     */
+    for (userp = ustartp + match, treep = tstartp + match; uendp - userp > WT_VECTOR_SIZE;
+         userp += WT_VECTOR_SIZE, treep += WT_VECTOR_SIZE) {
+        memcpy(&udata, userp, WT_VECTOR_SIZE);
+        memcpy(&tdata, treep, WT_VECTOR_SIZE);
+        if (udata.a != tdata.a || udata.b != tdata.b) {
+            match = (size_t)(userp - ustartp);
+            goto final128;
+        }
+    }
+
+    /*
+     * Rewind until there is exactly 16 bytes left. We know we started with at least 16, so we are
+     * still in bound.
+     */
+    memcpy(&udata, uendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+    memcpy(&tdata, tendp - WT_VECTOR_SIZE, WT_VECTOR_SIZE);
+    match = (size_t)(uendp - ustartp) - WT_VECTOR_SIZE;
+
+final128:
+    firsteq = udata.a == tdata.a;
+    u64 = firsteq ? udata.b : udata.a;
+    t64 = firsteq ? tdata.b : tdata.a;
+    match += firsteq ? sizeof(uint64_t) : 0;
```

Comment:
This is a bit silly, but I just noticed that if you write this as `match += firsteq * sizeof(uint64_t)` (which is exactly equivalent), that convinces gcc to make this whole section branchless. See `.L6` in the right pane of https://www.godbolt.org/z/nGxEWoaor, which corresponds ot the `final128` label in the C code.